eclipse-theia/theia
# Deduplicate InMemoryTextResource

#### What it does

Fixes https://github.com/eclipse-theia/theia/issues/9191 and an unreported issue with Search-In-Workspace: when clicking a search-and-replace suggestion, the preview editor should reveal the line on which the suggested change occurs. Currently on `master`, only the first suggested change in a file will be revealed.

(Please let me know if these should be separate PRs or instead should be flattened into a single commit. I found the latter issue while testing the fix for issue 9191, and it was unsatisfying to submit a fix that still looked broken.)

#### How to test


Test Bulk-Edit:
1. Open ts/tsx file, click a function name and press F2 to rename it.
2. Enter new name and press Shift+Enter.
3. A Refactor Preview pane should open in bottom with suggested changes.
4. Click on one of the changes suggested in that pane.
5. A split preview editor should open displaying the diff. This preview uses a `mem-txt` scheme.

<img width="506" alt="bulk-edit" src="https://user-images.githubusercontent.com/20653241/119026627-04512c80-b96b-11eb-84fd-f787f7b6830b.png">

Test Search-In-Workspace:
1. Use the search-in-workspace widget. Click the Toggle Replace triangle and enter replacement text.
2. Click one of the suggested search-and-replace operations.
3. A split preview editor should open displaying the diff. This preview uses a 'mem-txt' scheme.
4. Additionally, the preview editor should now reveal the line of the suggested change.

<img width="662" alt="s-i-w" src="https://user-images.githubusercontent.com/20653241/119026653-0c10d100-b96b-11eb-9ffb-7a3767eb80f6.png">

#### Review checklist


#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
